### ISSUE TITLE
Windows 64 5.x builds crash when switching between editor (SETM) and file browser

### ISSUE THREAD

**Issue by ChrisAshton84** [Beep6581/RawTherapee#3901]
Hi, first time RawTherapee user here - I'm on Win10-64b and crash both the currently available version 5 builds available when I enter the editor.  This is editing Canon CR2 files in a folder w/ ~2800 photos.  Host is a Nehalem i7 w/ 24GB ram and a nVidia 970.  The issue happens after navigating into the directory w/ the photos and double clicking on any of them.  The editor comes up and seems to crash the moment the photo is rendered.

Stepping back to the latest 4.x build gets into the editor successfully.

> Complete version information of the RawTherapee build you're using:

```
Version: 5.0-r1-gtk3
Branch: 5.0-r1-gtk3
Commit: 50114c1
Commit date: 2017-02-01
Compiler: gcc 5.3.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.18.0
Build type: release
Build flags: -m64 -mwin32 -mthreads -std=gnu++11 -mtune=generic -Werror=unused-label -fopenmp -Werror=unknown-pragmas -msse -msse2 -mwindows -fopenmp -Wno-aggressive-loop-optimizations -DNDEBUG -O3
Link flags: -m64 -mwin32 -mthreads -static-libgcc -mtune=generic -mwindows -s -O3
OpenMP support: ON
MMAP support: ON
```

> System information:

Windows 10 Pro
Core i7 860 @ 2.8 Ghz
24GB RAM (16 useable)
64-bit

> Tell us where you downloaded the RawTherapee build from:

http://rawtherapee.com/releases_head/windows/RawTherapee_5.0-r1-gtk3_WinVista_64.zip
http://rawtherapee.com/releases_head/windows/RawTherapee_5.1_WinVista_64.zip

> Explain the exact steps to reproduce the problem:

Browse using the file tree on the left column to the folder w/ my Canon .CR2 images.  Wait for images to fully load (2700-2800 images, takes a bit).  After loading, double click any image to open the editor.  Waits a moment, then once it renders the image it gets a C++ Runtime Error.

I can make images available if need be, but it seems to be any image I've tried.

> Show a screenshot of the problem, you can upload them to www.imgur.com:

http://imgur.com/PQ8NtI8

> Search the forum and our issue tracker before filing a new report as chances are that someone has already reported the problem before you, and duplication wastes time.
Make sure you use the latest version of RawTherapee as it's likely that a bug in an old version has been fixed in the latest one: Also happens exactly the same w/ latest version, listed above as another URL

> If your issue involves a crash, provide a stack backtrace if you can:

Error first reports

```
_create_dc_and_bitmap: The parameter is incorrect.
terminate called after throwing an instance of 'std::bad_alloc'
what() : std::bad_alloc
```
Many threads then exit with code 3...

The commands to get a bt didn't do anything for some reason.. no log.txt was created (I didn't realize until I exited).

**Comment by Beep6581** [comment]
Hi @ChrisAshton84 

If you're having trouble automatically creating a log file, you can do it by copying and pasting the backtrace. The 5.1 installer should contain a `rawtherapee-debug.exe` file. Run it through GDB as described [here](http://rawpedia.rawtherapee.com/How_to_write_useful_bug_reports#Step_by_step), then when RawTherapee crashes type "bt full" and paste everything which subsequently appears here into GitHub.

P.S. Please make sure you're using 5.1, nothing older than that.

**Comment by heckflosse** [comment]
Looks similar to #3885

**Comment by ChrisAshton84** [comment]
This happens identically w/ both 5.x releases available on the website.  I'll try again tonight to get a backtrace w/ the 5.1 release.  I don't think this is happening due to switching during thumbnail generation, btw, since I can wait an extended period of time, scan up and down the directory and see all thumbnails are visible, and then whenever I enter the editor it crashes (I believe it will generate the same crash whether double-clicking a thumbnail, right-clicking and choosing open, or clicking on the editor button).

**Comment by Beep6581** [comment]
@ChrisAshton84 does it crash if you copy one of the files you tried opening into an otherwise empty folder and opening it from there? If yes, please upload it using https://filebin.net/

**Comment by ChrisAshton84** [comment]
No log.txt was created because I was running in Program Files and my user didn't have permissions.  I re-ran gdb as Administrator (I trust you guys, I guess?) and, after the incredible load time, triggered the crash, and yet the backtrace still wouldn't generate.  All threads had exited.  gdb wasn't breaking on the exception.  I don't usually debug c++ but thought adding catch catch and catch throw would work.  No dice.  Any tips?  I'm trying a run w/ breaking on 'std::bad_alloc::bad_alloc()' as well, but it takes so long to start up...

**Comment by ChrisAshton84** [comment]
EDIT: Doing below doesn't trigger a crash if I don't have it running under gdb.exe / catch catch / catch throw so perhaps you're using exceptions for expected behavior and this is a red herring?  And, I'm able to bring up the editor when I'm in the directory w/ only one image.  But switching back to the large directory causes another runtime exception.

L O L

I created another folder as suggested w/ just a single image which I'd seen the crash occur w/.  After getting everything loaded up I double clicked that folder to change from the folder w/ 2800 photos.  It triggered a different crash.  I did have logging enabled and managed to get a bt for that thread (this time it caught the exception).  Then I tried to do the all threads bt and it actually crashed gdb.  Not kidding, the terminal I was running in is still open, but I got a Windows alert that gdb.exe had crashed and the terminal ends with:

```
#44 0x0000000063ea36c2 in ?? () from C:\Program Files\RawTherapee\5.1\libgtkmm-3.0-1.dll
No symbol table info available.
#45 0x00000000005a5565 in main (argc=<optimized out>, argv=<optimized out>) at Y:/RTSOURCE/rawtherapee/rtgui/main.cc:386
        exname = "C:\\Program Files\\RawTherapee\\5.1\\rawtherapee_debug.exe", '\000' <repeats 457 times>
        exePath = {static npos = 18446744073709551615, string_ = {static npos = 18446744073709551615, _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0xa869840 "C:\\Program Files\\RawTherapee\\5.1"}, _M_string_length = 32, {_M_local_buf = " \000\000\000\000\000\000\000\027\000\000\000\000\000\000", _M_allocated_capacity = 32}}}
        exnameU = L<error reading variable>
        consoleOpened = false
        m = <incomplete type>
        icon_path = {static npos = 18446744073709551615, string_ = {static npos = 18446744073709551615, _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0xcb7aab0 "C:\\Program Files\\RawTherapee\\5.1\\.\\images"}, _M_string_length = 41, {_M_local_buf = ")\000\000\000\000\000\000\000-\200_\017o\177\000", _M_allocated_capacity = 41}}}
        defaultIconTheme = {pCppObject_ = 0xcb7f210}
        screen = {pCppObject_ = 0xe60ecc0}
        rtWindow = 0xb869e40
```

The complete log including up to this point (I believe) is attached:
[log.txt](https://github.com/Beep6581/RawTherapee/files/1043480/log.txt)

**Comment by heckflosse** [comment]
I can reproduce the bug now this way:

I created a new folder with 1536 copies of amsterdam.pef
I opened the folder in rt an let the thumbs build
I closed rt

Now the part which is reproducible:

start rt
open the folder with the 1536 files
open the first file
open the second file from setm filmstrip
switch to file browser => crash

**Comment by heckflosse** [comment]
@ChrisAshton84 Are you running rt in Single Editor Tab Mode?

I get the crash only in Single Editor Tab Mode. In Multiple Editor Tabs Mode it does not crash.

Still searching...

**Comment by heckflosse** [comment]
An even easier way to reproduce:

Start rt (using SETM)
Open folder with 1536 files.
Switch to (empty) Editor
Switch back to file browser => crash

**Comment by heckflosse** [comment]
@Beep6581 @agriggio @Floessie 

Can you please try to reproduce the crash on Linux. The way to reproduce the crash I described in the posts above works 100% on Windows.

**Comment by heckflosse** [comment]
With this patch it doesn't crash:

```
diff --git a/rtgui/rtwindow.cc b/rtgui/rtwindow.cc
index ccf0b26e..596e6f0d 100644
--- a/rtgui/rtwindow.cc
+++ b/rtgui/rtwindow.cc
@@ -353,7 +353,7 @@ void RTWindow::on_mainNB_switch_page(Gtk::Widget* widget, guint page_num)
     if(!on_delete_has_run) {
         if(isEditorPanel(page_num)) {
             if (isSingleTabMode() && epanel) {
-                MoveFileBrowserToEditor();
+//                MoveFileBrowserToEditor();
             }
 
             EditorPanel *ep = static_cast<EditorPanel*>(mainNB->get_nth_page(page_num));
@@ -372,7 +372,7 @@ void RTWindow::on_mainNB_switch_page(Gtk::Widget* widget, guint page_num)
 
                 // Moving the FileBrowser only if the user has switched to the FileBrowser tab
                 if (mainNB->get_nth_page(page_num) == fpanel) {
-                    MoveFileBrowserToMain();
+//                    MoveFileBrowserToMain();
                 }
             }
         }

```

Of course we can't do that because it disables the filmstrip in SETM, but maybe it's a hint to find the culprit...

**Comment by Floessie** [comment]
@heckflosse Ingo, I'm sorry, but it doesn't crash here (Debian 9, today's release build, SETM-VT). :disappointed:

**Comment by Floessie** [comment]
@heckflosse At first glance, [this](https://github.com/Beep6581/RawTherapee/blob/dev/rtgui/rtwindow.cc#L744) looks interesting.

**Comment by heckflosse** [comment]
@Floessie I tried your suggestion but unfortunately it still crashes :disappointed:

**Comment by Floessie** [comment]
@heckflosse Ingo, I've no idea...

**Comment by heckflosse** [comment]
@Floessie Just to be sure: Did you try with a folder of 1536 (or more) copies of amsterdam.pef ?

**Comment by Floessie** [comment]
@heckflosse Sorry, no, I didn't realize that it **has** to be amsterdam.pef. I took exactly 1536 copies of an 18MP landscape CR2.

**Comment by heckflosse** [comment]
@Floessie No, it has not to be amsterdam.pef. But I just detected that I get the crash only when thumb infos are enabled.
Did you test with or without thumb infos?

**Comment by heckflosse** [comment]
@Floessie Next info: It also does not crash when thumb info is enabled but `Show basic Exif info` and `Show date and time` are disabled. Maybe a bug in this part of the code?
https://github.com/Beep6581/RawTherapee/blob/dev/rtgui/thumbbrowserentrybase.cc#L260

**Comment by heckflosse** [comment]
When I reduce font size from 8 to 7 it does not crash.
There is something wrong with the calculation of thumbnail height.
If I apply this patch I don't get a crash when switching between editor and file browser.
But it still crashes when reducing thumbnail size using zoom out (with and without patch).

```
diff --git a/rtgui/thumbbrowserentrybase.cc b/rtgui/thumbbrowserentrybase.cc
index 6fa8eac5..3bef29de 100644
--- a/rtgui/thumbbrowserentrybase.cc
+++ b/rtgui/thumbbrowserentrybase.cc
@@ -353,7 +353,7 @@ void ThumbBrowserEntryBase::resize (int h)
 {
     MYWRITERLOCK(l, lockRW);
 
-    height = h;
+    height = h + 20;
     int old_preh = preh, old_width = width;
 
     // dimensions of the button set

```

**Comment by heckflosse** [comment]
@Floessie @agriggio @Beep6581 Don't waste your time on this issue. I think I'm on to right trap now. With a bit of luck, I can solve it this weekend.

Ingo

**Comment by heckflosse** [comment]
This patch fixes the crashes when switching between file browser and editor in Single Editor Tab Mode in all my test cases:

```
diff --git a/rtgui/thumbbrowserentrybase.cc b/rtgui/thumbbrowserentrybase.cc
index 6fa8eac5..a6d9c1ea 100644
--- a/rtgui/thumbbrowserentrybase.cc
+++ b/rtgui/thumbbrowserentrybase.cc
@@ -314,33 +314,25 @@ void ThumbBrowserEntryBase::getTextSizes (int& infow, int& infoh)
     infow = 0;
 
     if (withFilename == WFNAME_FULL) {
-        // datetime
         fontd.set_weight (Pango::WEIGHT_NORMAL);
         context->set_font_description (fontd);
-        fn = w->create_pango_layout (datetimeline);
-        fn->get_pixel_size (dtlabw, dtlabh);
 
-        // basic exif data
-        fn = w->create_pango_layout (exifline);
-        fn->get_pixel_size (exlabw, exlabh);
 
         // add date/tile size:
         if (options.fbShowDateTime) {
+            // datetime
+            fn = w->create_pango_layout (datetimeline);
+            fn->get_pixel_size (dtlabw, dtlabh);
             infoh += dtlabh;
-
-            if (dtlabw + 2 * sideMargin > infow) {
-                infow = dtlabw + 2 * sideMargin;
-            }
         } else {
             dtlabw = dtlabh = 0;
         }
 
         if (options.fbShowBasicExif) {
+            // basic exif data
+            fn = w->create_pango_layout (exifline);
+            fn->get_pixel_size (exlabw, exlabh);
             infoh += exlabh;
-
-            if (exlabw + 2 * sideMargin > infow) {
-                infow = exlabw + 2 * sideMargin;
-            }
         } else {
             exlabw = exlabh = 0;
         }
@@ -401,20 +393,7 @@ void ThumbBrowserEntryBase::resize (int h)
 
     calcThumbnailSize ();  // recalculates prew
 
-    width = prew + 2 * sideMargin + 2 * borderWidth;
-
-    if (    (parent->getLocation() != ThumbBrowserBase::THLOC_EDITOR && options.showFileNames && !options.overlayedFileNames)
-            || (parent->getLocation() == ThumbBrowserBase::THLOC_EDITOR && options.filmStripShowFileNames && !options.filmStripOverlayedFileNames)) {
-        width = prew + 2 * sideMargin + 2 * borderWidth;
-
-        if (width < infow + 2 * sideMargin + 2 * borderWidth) {
-            width = infow + 2 * sideMargin + 2 * borderWidth;
-        }
-    }
-
-    if (width < bsw + 2 * sideMargin + 2 * borderWidth) {
-        width = bsw + 2 * sideMargin + 2 * borderWidth;
-    }
+    width = std::max(prew, bsw) + 2 * sideMargin + 2 * borderWidth;
 
     if ( preh != old_preh || width != old_width ) {
         delete [] preview;

```

@gaaned92 Can you please make a windows build with that patch? I would be interested to know whether it fixes the crashes reported by @ChrisAshton84 as well.

**Comment by heckflosse** [comment]
I found another constellation which crashes even without switching from file browser to SETM editor. It just crashes by switching folders in that special constellation. I'll investigate...

**Comment by gaaned92** [comment]
Edit:
RawTherapee_dev_5.1-38-g982b291a_patched_WinVista_64.zip uploaded at [https://drive.google.com/open?id=0B2q9OrgyDEfPS2FpdDAtMVI1RG8](https://drive.google.com/open?id=0B2q9OrgyDEfPS2FpdDAtMVI1RG8)

**Comment by heckflosse** [comment]
@gaaned92 Thanks for the build :+1:

**Comment by ChrisAshton84** [comment]
@gaaned92 Thanks, I'll try to test this tonight.  I was on vacation this weekend have another 1000 photos to test 👍

**Comment by heckflosse** [comment]
@ChrisAshton84 If it still crashes, please try also Multiple Editor Tabs Mode

**Comment by ChrisAshton84** [comment]
This still crashed at roughly the same point - without gdb, when trying to go to the editor (after the images all loaded their thumbnails), the gui started bugging out.  Most of the window was completely blank, and even the min / max / close buttons stopped drawing correctly when I moused over.  It then closed.  Previously it had at least shown all the gui components as it started to load the editor for that image, then crashed.
In gdb, this time it crashed immediately after loading all the thumbnails, without trying to enter the editor.  I still see the 'terminate called after throwing an instance of 'std::bad_alloc'' message, with no useful backtrace in any thread (all threads have exited, and no stack remains):

bt
No stack.
info threads
No threads.

Do I need some other debug libraries installed before this will work?  I don't have any dev environment on my Windows machine.

**Comment by ChrisAshton84** [comment]
@heckflosse I tried METM and it crashed the gui as soon as I tried to open a photo for editing.  Gui didn't close but stopped redrawing anything.  I can try gdb if you can help me figure out how to get useful backtraces or breakpoints for you.  Thanks!

**Comment by gaaned92** [comment]
@ChrisAshton84 
The gdb.exe debugger and debugexe rawtherapee_debug.exe are packed into the build so, to get a backtrace:
- open a command window
- go to the rawtherapee installation directory
- type _gdb rawtherapee_debug.exe_ . you get a gdb prompt then type _run_
- rawtherapee window opens. Do what you want in rawtherapee window
- when it crashes or hangs go back to command window. You have the gdb> prompt . Type _bt full_
- copy and paste bt message here. Use "insert code" to make it readable.

You can have a look at [http://rawpedia.rawtherapee.com/How_to_write_useful_bug_reports](http://rawpedia.rawtherapee.com/How_to_write_useful_bug_reports) to generate a log file.

**Comment by Floessie** [comment]
@ChrisAshton84 
> In gdb, this time it crashed immediately after loading all the thumbnails, without trying to enter the editor. I still see the 'terminate called after throwing an instance of 'std::bad_alloc'' message, with no useful backtrace in any thread (all threads have exited, and no stack remains)

Yep, that's too late, the milk is spilt. Maybe you can catch it with [`catch throw`](https://stackoverflow.com/questions/6835728/gdb-how-to-break-when-a-specific-exception-type-is-thrown), but afaik @heckflosse had problems with it under Windows. Maybe you'll have better luck...

HTH,
Flössie

**Comment by ChrisAshton84** [comment]
I've tried both 'gdb rawtherapee_debug.exe' and inside gdb, using the 'file' command to load the debug executable.  I also tried catch throw and catch catch.  I posted the results earlier; taking the backtrace at that point (which wasn't the bad_alloc error, IIRC) caused gdb to crash!

**Comment by Floessie** [comment]
@ChrisAshton84 Sorry, now I remember I've already read your `log.txt` from above where you state what tricks you did with GDB. I didn't intent to treat you like a noob and should have read the whole thread before giving advices... :neutral_face:

**Comment by ChrisAshton84** [comment]
No worries.  I realize there's something off in my case.  gdb shouldn't crash, and that likely isn't rt's issue (though how it packages gdb, where its gdb came from, etc might be a problem for me apparently).

**Comment by ChrisAshton84** [comment]
And, while I'm excellent at debugging low level C on Linux, I can't say as much for debugging gui toolkits, C++ exceptions, and have no experience whatsoever on Windows in the last 10 years.  It just happens to be my main gaming machine's OS which is where I'd like to run RT.

**Comment by heckflosse** [comment]
@ChrisAshton84 

Please take a look at https://github.com/Beep6581/RawTherapee/issues/3937#issuecomment-310196217
Does setting the limit to 20000 solve the issue?
I know that increasing the limit is a workaround and not a solution, but currently each thumb needs 2 GDI handles. Additionally the handles are leaking when switching folders in rt (of course we have to solve this).

**Comment by heckflosse** [comment]
Fixed with e4c75c4
